orbit/pkg/luks/luks_linux.go

```diff
@@ -39,39 +38,85 @@ const (
 	userKeySlot          = 0 // Key slot 0 is assumed to be the location of the user's passphrase
 )
 
-var ErrKeySlotFull = regexp.MustCompile(`Key slot \d+ is full`)
+var errKeySlotNotFound = errors.New("key slot not found")
 
-func isInstalled(toolName string) bool {
-	path, err := exec.LookPath(toolName)
+func withRootDevicePath(fn func(string) error) error {
+	devicePath, err := lvm.FindRootDisk()
 	if err != nil {
-		return false
+		return fmt.Errorf("failed to find LUKS Root Partition: %w", err)
 	}
-	return path != ""
+	return fn(devicePath)
 }
 
 func (lr *LuksRunner) Run(oc *fleet.OrbitConfig) error {
 	ctx := context.Background()
 
-	if !oc.Notifications.RunDiskEncryptionEscrow {
+	if oc.Notifications.RunDiskEncryptionEscrow {
+		return checkInstalled([]string{"cryptsetup"}, func() error {
+			return withRootDevicePath(func(devicePath string) error {
+				return lr.newUserKeyWorkflow(ctx, devicePath)
+			})
+		})
+	}
+
+	salt := oc.Notifications.EscrowedKeySalt
+	slot := oc.Notifications.EscrowedKeySlot
+	if len(salt) == 0 || slot == nil {
+		log.Debug().Msg("No salt or key slot provided, skipping escrowed key validation")
 		return nil
 	}
+	return checkInstalled([]string{"cryptsetup"}, func() error {
```

Comment:
I would add something similar to:

https://github.com/fleetdm/fleet/blob/94aa2004242e5fec3676930bbf316148d52eeff4/orbit/pkg/update/escrow_buddy.go#L57-L60

so that this check doesn't run every 30s, it's probably better to run this at 1h interval (similar to our refetch interval).

Reply:
Oh nice, was not aware of that - will do the refactor.